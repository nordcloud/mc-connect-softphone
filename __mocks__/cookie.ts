export default {
  serialize: jest.fn().mockReturnValue('dummyCookieSerializeValue'),
};
