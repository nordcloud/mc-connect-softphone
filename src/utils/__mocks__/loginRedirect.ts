export const loginRedirect = jest.fn().mockReturnValue('dummyLoginRedirect');
