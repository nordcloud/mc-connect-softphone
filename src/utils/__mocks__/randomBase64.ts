export const randomBase64 = jest.fn().mockResolvedValue('randomBase64Value');
