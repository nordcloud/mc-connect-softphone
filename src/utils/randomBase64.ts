import crypto from 'crypto';
import { promisify } from 'util';

export async function randomBase64(size: number) {
  const cryptoRandomBytes = promisify(crypto.randomBytes);
  const bytes = await cryptoRandomBytes(size);

  return bytes.toString('base64url');
}
