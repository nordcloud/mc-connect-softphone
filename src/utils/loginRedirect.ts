import cookie from 'cookie';
import crypto from 'crypto';
import { StatusCodes } from 'http-status-codes';

import { LOGIN_FLOW_COOKIE } from '../consts';
import { getStage } from './getStage';
import { getStageConsts } from './getStageConsts';
import { randomBase64 } from './randomBase64';

const { OAUTH_BASE_URL, OAUTH_CLIENT_ID } = getStageConsts();

export function loginRedirect() {
  return oauthRedirect();
}

async function oauthRedirect() {
  const { LOGIN_CALLABCK_URL } = process.env;

  if (!LOGIN_CALLABCK_URL) {
    throw new Error('Missing LOGIN_CALLABCK_URL');
  }

  const oauthUrl = new URL(`${OAUTH_BASE_URL}/authorize`);

  const stateNonce = await randomBase64(128);
  const idTokenNonce = await randomBase64(128);
  const codeVerifier = await randomBase64(96);
  const codeChallenge = crypto
    .createHash('sha256')
    .update(codeVerifier)
    .digest('base64url');

  const params = new URLSearchParams({
    client_id: OAUTH_CLIENT_ID,
    code_challenge: codeChallenge,
    code_challenge_method: 'S256',
    nonce: idTokenNonce,
    redirect_uri: LOGIN_CALLABCK_URL,
    response_type: 'code',
    scope: 'openid profile email',
    state: stateNonce,
  });

  oauthUrl.search = params.toString();

  const loginFlowCookie = cookie.serialize(
    LOGIN_FLOW_COOKIE,
    JSON.stringify({
      stateNonce,
      idTokenNonce,
      codeVerifier,
    }),
    {
      httpOnly: true,
      secure: getStage() !== 'local',
      sameSite: 'lax',
    }
  );

  console.log(`Redirecting to OAuth ${oauthUrl.href}`);

  return {
    statusCode: StatusCodes.MOVED_TEMPORARILY,
    headers: { location: oauthUrl.href },
    cookies: [loginFlowCookie],
  };
}
