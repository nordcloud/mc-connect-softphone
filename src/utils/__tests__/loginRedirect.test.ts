import cookie from 'cookie';

import { loginRedirect } from '../loginRedirect';
import { randomBase64 } from '../randomBase64';

jest.mock('../randomBase64');

beforeEach(() => {
  process.env.LOGIN_CALLABCK_URL = 'dummyLoginCallbackUrl';
});

describe('loginRedirect', () => {
  it('calls randomBase64 with correct args', async () => {
    await loginRedirect();
    expect(jest.mocked(randomBase64).mock.calls).toMatchSnapshot();
  });

  // it('calls crypto.createHash with correct args', async () => {
  //   await loginRedirect();
  //   expect(jest.mocked(crypto.createHash).mock.calls).toMatchSnapshot();
  // });

  // it('calls crypto.createHash().update with correct args', async () => {
  //   await loginRedirect();
  //   expect(jest.mocked(crypto.createHash('').update).mock.calls).toMatchSnapshot();
  // });

  // it('calls crypto.createHash().update().digest with correct args', async () => {
  //   await loginRedirect();
  //   expect(
  //     jest.mocked(crypto.createHash('').update('').digest).mock.calls
  //   ).toMatchSnapshot();
  // });

  it('calls cookie.serialize with correct args', async () => {
    await loginRedirect();
    expect(jest.mocked(cookie.serialize).mock.calls).toMatchSnapshot();
  });

  it('resolves with a correct value', () => {
    expect(loginRedirect()).resolves.toMatchSnapshot();
  });

  it('rejects with a correct value when LOGIN_CALLABCK_URL is missing', () => {
    delete process.env.LOGIN_CALLABCK_URL;
    expect(loginRedirect()).rejects.toMatchSnapshot();
  });
});
