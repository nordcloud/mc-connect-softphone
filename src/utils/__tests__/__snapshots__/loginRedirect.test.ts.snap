// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`loginRedirect calls cookie.serialize with correct args 1`] = `
[
  [
    "loginFlowState",
    "{"stateNonce":"randomBase64Value","idTokenNonce":"randomBase64Value","codeVerifier":"randomBase64Value"}",
    {
      "httpOnly": true,
      "sameSite": "lax",
      "secure": true,
    },
  ],
]
`;

exports[`loginRedirect calls randomBase64 with correct args 1`] = `
[
  [
    128,
  ],
  [
    128,
  ],
  [
    96,
  ],
]
`;

exports[`loginRedirect rejects with a correct value when LOGIN_CALLABCK_URL is missing 1`] = `[Error: Missing LOGIN_CALLABCK_URL]`;

exports[`loginRedirect resolves with a correct value 1`] = `
{
  "cookies": [
    "dummyCookieSerializeValue",
  ],
  "headers": {
    "location": "https://nordcloud.okta.com/oauth2/v1/authorize?client_id=0oaqyrebtuY1eagfS357&code_challenge=mGTxtN-qAtITFzktbpDTLspPVhDJXlUJyOIIjyDopd0&code_challenge_method=S256&nonce=randomBase64Value&redirect_uri=dummyLoginCallbackUrl&response_type=code&scope=openid+profile+email&state=randomBase64Value",
  },
  "statusCode": 302,
}
`;
